### ISSUE TITLE
allow end users to customize module sharing via federation

### ISSUE THREAD

**Issue by tmpaul** [webpack/webpack#17404]
## Feature request

**What is the expected behavior?**

Developers should be able to customize the sharing behavior for a dependency based off where's it's being loaded in, or what version of the dependency is being shared. This issue is the first in line for the companion [PR here](https://github.com/webpack/webpack/pull/17396)

**What is motivation or use case for adding/changing the behavior?**

In standard module federation implementation, developers can specify that a given dependency is "shared" for consumption, i.e other consumers can remotely consume a shared module. They do this by adding the module to `shared` property in `ModuleFederationPlugin`. However, developers are not allowed to customize how this sharing works with a complex nested dependency tree.

Example repository showcasing the problem: [Different major versions of a shared dependency d3-array.](https://github.com/tmpaul/module-federation-enhancements-showcase/tree/main/exclusion#different-major-versions-of-a-shared-dependency-d3-array).
   * Clone, `yarn` + `yarn serve` for production builds, `yarn start` for development builds

In this example repo, there are two versions of `d3-array` (`3.x` and `1.x`). By marking `d3-array` as shared, both versions are automatically marked for shared consumption. In many cases, this is undesirable. Older versions may contain non-treeshakeable code, buggy code, or obsolete code. The main use-case involves sharing "top" level dependencies present in the root `package.json` file which is what developers have control over. They do not directly control how another version of the same package is brought in transitively. 

> Today, there is no way to tell `webpack`: "Please don't share a package when it is requested in a specific context `x` or if the major version is `y`, but only share newer versions".

The problem then is that the older version is also built out as a shared chunk, preventing tree-shaking. The bundle-size with exclusion in effect (i.e excluding 1.x from sharing) is lower than standard webpack sharing. In a real life project with tons of nested dependencies, this number can be staggeringly high. In our company, I have seen remoteEntry files ~ 200kb due to tons of nested dependency sharing.

The following table shows the bundle size for not sharing vs sharing all versions of a package vs sharing only
a specific major version of a package

| Sharing Type | JS size |  Versions |
---------------|-----------|---------|
| Not shared |   92 kb |  N/A |
| Not excluded |  101 kb     | 3.2.4,1.2.4 |
| Excluded  |    98.5 kb   | 3.2.4 |

**How should this be implemented in your opinion?**

Developers must be able to specify when to exclude a given dependency from sharing. This can come in a few varieties:
a. Exclude from sharing when imported by a file in a specific directory (`context` or `issuer` in webpack terms).
b. Exclude from sharing when a specific version or range criteria is satisfied.

To this end, I can think of two ways to implement this:

```javascript
shared: {
   "d3-array": {
        "exclusionCriteria": {
              /*To match issuer or context. For example, if d3-array is imported by d3-scale, exclude, */
              "context": string | RegExp | undefined | (context: string) => boolean
              /*To match a specific version or range */
              "version": "1.2.4" | "^1.0.0"
         }
    }
}
```

I added the `context` behavior in the PR, but in all honesty, I think the `version` or `range` based exclusion is far more meaningful because the `context` can be confusing in monorepo setups like `lerna`

**Open Questions**
1. How do we handle the scenario where a package is excluded based off a specific `context`, but is allowed to be shared in another context with a matching version ?

     For instance, if `d3-array@1.2.4` is excluded for all imports from `d3-scale`, but is not excluded for any other contexts, and another context pulls in `d3-array@1.2.4`, then excluding from sharing is actually detrimental. Because, there is now a shared chunk@1.2.4 and a tree-shaken bundled in version@1.2.4. This can be solved by restricting the exclusion criteria to just version or range.

**Are you willing to work on this yourself?**
yes

Related issue: https://github.com/webpack/webpack/issues/15971

**Comment by alexander-akait** [comment]
I see the problem

>   /*To match issuer or context. For example, if d3-array is imported by d3-scale, exclude, */
              "context": string | RegExp | undefined
              
I don't like it, because some modules can be without an issuer or can have multiple issuers, it is not flexible and can create a false sense that everything is set up correctly, we should use the module here, yes, you will required to review manually (or using bundle analizer to catch some cases)

**Comment by tmpaul** [comment]
> I don't like it, because some modules can be without an issuer or can have multiple issuers

I agree that it can be confusing. In practice, without the full stack of issuer calls, it is hard to determine whether to exclude or not. But do you agree with the version range exclusion proposal instead ? I feel that it is a valid use-case, especially given that today there is no control over the sharing behavior. I can adjust the PR to use that instead of `context`

**Comment by alexander-akait** [comment]
> But do you agree with the version range exclusion proposal instead ? I feel that it is a valid use-case, especially given that today there is no control over the sharing behavior. I can adjust the PR to use that instead of context

Agree, it is useful, there is only one things - do you want to support something like `^4.2` (i.e. semver matching)?

**Comment by tmpaul** [comment]
> do you want to support something like ^4.2 (i.e. semver matching)?

Yes. Because most of the nested dependency sharing comes from ranges, rather than pinned versions. And since a pinned version is a special case for a range, handling a range should automatically take care of the pinned version. My thought process is as follows:

```
if (semver.intersectsRange(versionOfSharedPackageOrRange, userSuppliedExclusionRange) {
   // Exclude
}
// Else Include
```

The `semver.intersectsRange` is a pseduo function, I will most likely use the inbuilt lib [webpack/lib/util/semver.js](https://github.com/webpack/webpack/blob/main/lib/util/semver.js)

**Comment by alexander-akait** [comment]
Yeah, sounds good for me, I think we can split this PR in two parts (we can?):
- `version`, it is already can be useful
- Detecting `context`  - we use this for externals - https://webpack.js.org/configuration/externals/#function, so technically we can do the same it will allow be flexible

**Comment by ErreMalote** [comment]
bump
